### ISSUE TITLE
GPG Encrypted mail does not display: This is an encrypted message and can not be displayed. Sorry!

### ISSUE THREAD

**Issue by SjoerdV** [roundcube/roundcubemail#9534]
### Prerequisites

- [X] I have [searched](https://github.com/roundcube/roundcubemail/issues?q=is%3Aissue) for duplicate or closed issues
- [X] I can recreate the issue with all plugins disabled

### Describe the issue

Several emails (i have attached one), gpg encrypted, will not display contents properly
Messages do show Subject, From, To, Date, Summary, Headers, but no Body
instead the error 'This is an encrypted message and can not be displayed. Sorry!' is displayed.

The Public and Private key is sure to be working because 
- other mails (encrypted with the same public key) are opening correctly with the same private key.
- Attached example does open correctly in Thunderbird 115.12.0

[test.eml.txt](https://github.com/user-attachments/files/16104618/test.eml.txt)

### What browser(s) are you seeing the problem on?

Firefox

### What version of PHP are you using?

8.2.20

### What version of Roundcube are you using?

1.6.7

### JavaScript errors

none in console

### PHP errors

nothing in syslog

**Comment by pabzm** [comment]
As far as I can tell the message structure is OK, but the ciphertext has only one anonymous recipient, so GnuPG doesn't know which key to decrypt with. Could that be the problem?

**Comment by pabzm** [comment]
You could also set `enigma_debug` in the plugin's config-file and provide the log (please make sure it doesn't contain sensitive data).

**Comment by SjoerdV** [comment]
@pabzm , thanks for your help! The mail in question does not come into the GPG subroutine, so I have no logs for that part.
The regular RC logs show this: `BODYSTRUCTURE (("text" "plain" ("charset" "us-ascii")` while a GPG encrypted mail that does process shows this: `BODYSTRUCTURE (("application" "pgp-encrypted" NIL NIL "PGP/MIME version identification" "7bit" 12 NIL NIL NIL NIL)`. So it seems the erroneous mail does not get detected with the right ContentType, while it clearly is the right ContentType.

<details><summary>RC log that FAILS to start processing the GPG payload, see attachment in OP</summary>

```log
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] Connecting to ssl://imap.mail.server.com:993...
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * OK [CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE UIDPLUS LITERAL+ AUTH=PLAIN AUTH=LOGIN AUTH=OAUTHBEARER AUTH=XOAUTH2] Dovecot ready.
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] C: A0001 AUTHENTICATE LOGIN
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: + VXNlcm5hbWU6
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] C: ****** [38]
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: + UGFzc3dvcmQ6
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] C: ****** [42]
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: A0001 OK [CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS THREAD=ORDEREDSUBJECT MULTIAPPEND URL-PARTIAL CATENATE UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS BINARY MOVE SNIPPET=FUZZY PREVIEW=FUZZY PREVIEW STATUS=SIZE SAVEDATE SEARCH=X-MIMEPART XDOVECOT ACL UIDPLUS LITERAL+ NOTIFY SPECIAL-USE FILTER=SIEVE COMPRESS=DEFLATE XAPPLEPUSHSERVICE QUOTA ACL RIGHTS=texk] Logged in
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] C: A0002 ID ("name" "Roundcube" "version" "1.6.7" "php" "8.2.20" "os" "Linux" "command" "/mail/?_task=mail&_caps=pdf%3D1%2Cflash%3D0%2Ctiff%3D0%2Cwebp%3D1%2Cpgpmime%3D0&_uid=47&_mbox=Drafts&_framed=1&_action=preview")
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * ID ("name" "Dovecot")
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: A0002 OK ID completed (0.001 + 0.000 secs).
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] C: A0003 SELECT Drafts
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * FLAGS (\Answered \Flagged \Deleted \Seen \Draft)
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * OK [PERMANENTFLAGS (\Answered \Flagged \Deleted \Seen \Draft \*)] Flags permitted.
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * 6 EXISTS
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * 0 RECENT
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * OK [UIDVALIDITY 1719958355] UIDs valid
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * OK [UIDNEXT 48] Predicted next UID
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * OK [HIGHESTMODSEQ 124] Highest
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: A0003 OK [READ-WRITE] Select completed (0.001 + 0.000 secs).
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] C: A0004 UID FETCH 47 (UID RFC822.SIZE FLAGS INTERNALDATE BODYSTRUCTURE BODY.PEEK[HEADER.FIELDS (DATE FROM TO SUBJECT CONTENT-TYPE CC REPLY-TO LIST-POST DISPOSITION-NOTIFICATION-TO X-PRIORITY CONTENT-TRANSFER-ENCODING BCC IN-REPLY-TO MAIL-FOLLOWUP-TO MAIL-REPLY-TO MESSAGE-ID REFERENCES RESENT-BCC RETURN-PATH SENDER X-DRAFT-INFO)])
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: * 6 FETCH (UID 47 RFC822.SIZE 4318 FLAGS (\Seen) INTERNALDATE "08-Jul-2024 22:14:47 +0200" BODYSTRUCTURE (("text" "plain" ("charset" "us-ascii") NIL NIL "7bit" 0 0 NIL NIL NIL NIL) "encrypted" ("protocol" "application/pgp-encrypted" "boundary" "--------------QGBFJzxTf6SA80kyYAzDzAyE") NIL NIL NIL) BODY[HEADER.FIELDS (DATE FROM TO SUBJECT CONTENT-TYPE CC REPLY-TO LIST-POST DISPOSITION-NOTIFICATION-TO X-PRIORITY CONTENT-TRANSFER-ENCODING BCC IN-REPLY-TO MAIL-FOLLOWUP-TO MAIL-REPLY-TO MESSAGE-ID REFERENCES RESENT-BCC RETURN-PATH SENDER X-DRAFT-INFO)] {283}
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: From: replaced@replaced.com\x0d
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: To: <replaced@replaced.com>\x0d
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: Subject: =?UTF-8?B?U2VjdXJlLCBncmVlbiAmIGFkLWZyZWUh?=\x0d
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: Date: Fri, 15 Mar 2024 18:09:09 +0000\x0d
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: Content-Type: multipart/encrypted; \x0d
Jul 08 22:25:12 MAILSERVER01 roundcube[771]:  protocol="application/pgp-encrypted";\x0d
Jul 08 22:25:12 MAILSERVER01 roundcube[771]:  boundary="--------------QGBFJzxTf6SA80kyYAzDzAyE"\x0d
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: \x0d
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: 
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: )
Jul 08 22:25:12 MAILSERVER01 roundcube[771]: <kicig4t9> [587F] S: A0004 OK Fetch completed (0.001 + 0.000 secs).
```

</details>

<details><summary>RC log that SUCCEEDS to start processing the GPG payload</summary>

```log
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] Connecting to ssl://imap.mail.server.com:993...
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * OK [CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE UIDPLUS LITERAL+ AUTH=PLAIN AUTH=LOGIN AUTH=OAUTHBEARER AUTH=XOAUTH2] Dovecot ready.
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] C: A0001 AUTHENTICATE LOGIN
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: + VXNlcm5hbWU6
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] C: ****** [38]
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: + UGFzc3dvcmQ6
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] C: ****** [42]
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: A0001 OK [CAPABILITY IMAP4rev1 SASL-IR LOGIN-REFERRALS ID ENABLE IDLE SORT SORT=DISPLAY THREAD=REFERENCES THREAD=REFS THREAD=ORDEREDSUBJECT MULTIAPPEND URL-PARTIAL CATENATE UNSELECT CHILDREN NAMESPACE UIDPLUS LIST-EXTENDED I18NLEVEL=1 CONDSTORE QRESYNC ESEARCH ESORT SEARCHRES WITHIN CONTEXT=SEARCH LIST-STATUS BINARY MOVE SNIPPET=FUZZY PREVIEW=FUZZY PREVIEW STATUS=SIZE SAVEDATE SEARCH=X-MIMEPART XDOVECOT ACL UIDPLUS LITERAL+ NOTIFY SPECIAL-USE FILTER=SIEVE COMPRESS=DEFLATE XAPPLEPUSHSERVICE QUOTA ACL RIGHTS=texk] Logged in
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] C: A0002 ID ("name" "Roundcube" "version" "1.6.7" "php" "8.2.20" "os" "Linux" "command" "/mail/?_task=mail&_caps=pdf%3D1%2Cflash%3D0%2Ctiff%3D0%2Cwebp%3D1%2Cpgpmime%3D0&_uid=18&_mbox=Drafts&_framed=1&_action=preview")
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * ID ("name" "Dovecot")
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: A0002 OK ID completed (0.001 + 0.000 secs).
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] C: A0003 SELECT Drafts
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * FLAGS (\Answered \Flagged \Deleted \Seen \Draft)
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * OK [PERMANENTFLAGS (\Answered \Flagged \Deleted \Seen \Draft \*)] Flags permitted.
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * 6 EXISTS
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * 0 RECENT
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * OK [UIDVALIDITY 1719958355] UIDs valid
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * OK [UIDNEXT 48] Predicted next UID
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * OK [HIGHESTMODSEQ 124] Highest
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: A0003 OK [READ-WRITE] Select completed (0.001 + 0.000 secs).
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] C: A0004 UID FETCH 18 (UID RFC822.SIZE FLAGS INTERNALDATE BODYSTRUCTURE BODY.PEEK[HEADER.FIELDS (DATE FROM TO SUBJECT CONTENT-TYPE CC REPLY-TO LIST-POST DISPOSITION-NOTIFICATION-TO X-PRIORITY CONTENT-TRANSFER-ENCODING BCC IN-REPLY-TO MAIL-FOLLOWUP-TO MAIL-REPLY-TO MESSAGE-ID REFERENCES RESENT-BCC RETURN-PATH SENDER X-DRAFT-INFO)])
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * 5 FETCH (UID 18 RFC822.SIZE 3980 FLAGS (\Seen \Draft) INTERNALDATE "04-Jul-2024 15:36:49 +0200" BODYSTRUCTURE (("application" "pgp-encrypted" NIL NIL "PGP/MIME version identification" "7bit" 12 NIL NIL NIL NIL)("application" "octet-stream" ("name" "encrypted.asc") NIL "OpenPGP encrypted message" "7bit" 2825 NIL ("inline" ("filename" "encrypted.asc")) NIL NIL) "encrypted" ("protocol" "application/pgp-encrypted" "boundary" "------------dO30Fb0GDw4HJkgp73LIB2QS") NIL ("en-US" "nl") NIL) BODY[HEADER.FIELDS (DATE FROM TO SUBJECT CONTENT-TYPE CC REPLY-TO LIST-POST DISPOSITION-NOTIFICATION-TO X-PRIORITY CONTENT-TRANSFER-ENCODING BCC IN-REPLY-TO MAIL-FOLLOWUP-TO MAIL-REPLY-TO MESSAGE-ID REFERENCES RESENT-BCC RETURN-PATH SENDER X-DRAFT-INFO)] {368}
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: Message-ID: <af8d20a6-5a21-47d8-a8a5-f061f3c7890c@replaced.com>\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: Date: Thu, 4 Jul 2024 15:36:48 +0200\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: From: My Name <replaced@replaced.com>\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: To: "My Name" <replaced@replaced.com>\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: Subject: ...\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: Content-Type: multipart/encrypted;\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]:  protocol="application/pgp-encrypted";\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]:  boundary="------------dO30Fb0GDw4HJkgp73LIB2QS"\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: \x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: 
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: )
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: A0004 OK Fetch completed (0.001 + 0.000 secs).
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] C: A0005 UID FETCH 18 (BODY.PEEK[2.MIME])
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * 5 FETCH (UID 18 BODY[2.MIME] {167}
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: Content-Type: application/octet-stream; name="encrypted.asc"\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: Content-Description: OpenPGP encrypted message\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: Content-Disposition: inline; filename="encrypted.asc"\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: \x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: 
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: )
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: A0005 OK Fetch completed (0.001 + 0.000 secs).
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] C: A0006 UID FETCH 18 (BODY.PEEK[2])
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: * 5 FETCH (UID 18 BODY[2] {2825}
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: -----BEGIN PGP MESSAGE-----\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: \x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: wV4DYDHSwVcnJGsSAQdAcKYLHN1p6RdQQ0ZxatuIFtqn59wUNb68z2R+WNwDNH8w5ijsYMgPid4h\x0d
...
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: Fz2/vQzChjPJOwiZWfCWPtUbJ8GvAOW6VQHfftDOvd7268CzcfL7sXu06zxQFKiXXfzGjbazGFZt\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: ZaT+DPHLiB9NhPoYiW3Y8oHk5YjoDh1Hn
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: +DdhSdV50YsoxCqzuno3s4cXanjHRhFAHELpfgZyDOg\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: MT4dY0JxP1/H8IrrsSkJao9URCIEbcXSltRv8HdTGUn5SvwXK3D6zXyvuOee94d4SmviGyuddsr5\x0d
...
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: /kxQ//hXUmlT94mQoO7PQkAIa5VD6R5VTRCAzZcfaeBHFDuMvemGQ9L3RoHfwC+8P8IbVsFoxJWA\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: GEETFrHxFm30BHtJjI3MkelK+gas\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: =JzA/\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: -----END PGP MESSAGE-----\x0d
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: 
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: )
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> [31CF] S: A0006 OK Fetch completed (0.001 + 0.000 secs).
Jul 08 22:27:55 MAILSERVER01 roundcube[773]: <kicig4t9> GPG: OPENING GPG SUBPROCESS WITH THE FOLLOWING COMMAND:
...
```

</details>

**Comment by alecpl** [comment]
So, it is a dovecot issue, but should we add a workaround? We already have some code that detects invalid BODYSTRUCTURE response and in such case decodes the whole message not relying on it. Maybe we should/could handle this case there too.

https://github.com/roundcube/roundcubemail/blob/6b7926c0356f970883bd2a1201adb7ab85fa8927/program/lib/Roundcube/rcube_imap.php#L1958

What dovecot version it is?

**Comment by SjoerdV** [comment]
Ah, dovecot. my version is: 2.3.19.1 (9b53102964)

Reviewing that piece of code, it clearly says not to rely on the `rcube_mime` routine because it is very badly optimized, so I would not recommend using it.

**Anyway... you guys (@alecpl @pabzm ) did make me look further into why this particular email was not rendering in Roundcube and I am afraid it was all my fault (well actually it was the Tuta mail export routine, but I should have seen it). The `boundary="..."` section of the eml file had 2 too many dashes...**

It seems Thunderbird processes the message completely independent of any standard, so it's much more forgiving than 'plain' dovecot. I have learned a lot and hope you will forgive me wasting your time a bit. 
In the end I used 'meld' to visually compare a correct mail and the faulty one and noticed the dashes. Hope this retrospect will help someone. Thanks again!
